src/py/flwr/common/backoff.py

```diff
@@ -0,0 +1,278 @@
+    Union,
+)
+
+
+def exponential(
+    base_delay: float = 1,
+    multiplier: float = 2,
+    max_delay: Optional[int] = None,
+) -> Generator[float, None, None]:
+    """Wait time generator for exponential backoff strategy.
+
+    Parameters
+    ----------
+    base_delay: float (default: 1)
+        Initial delay duration before the first retry.
+    multiplier: float (default: 2)
+        Factor by which the delay is multiplied after each retry.
+    max_delay: Optional[float] (default: None)
+        The maximum delay duration between two consecutive retries.
+    """
+    delay = base_delay if max_delay is None else min(base_delay, max_delay)
+    while True:
+        yield delay
+        delay *= multiplier
+        if max_delay is not None:
+            delay = min(delay, max_delay)
+
+
+def constant(
+    interval: Union[float, Iterable[float]] = 1,
+) -> Generator[float, None, None]:
+    """Wait time generator for specified intervals.
+
+    Parameters
+    ----------
+    interval: Union[float, Iterable[float]] (default: 1)
+        A constant value to yield or an iterable of such values.
+    """
+    if not isinstance(interval, Iterable):
+        interval = itertools.repeat(interval)
+    yield from interval
+
+
+def random_jitter(value: float) -> float:
+    """Jitter the value by a random number of milliseconds.
+
+    This adds up to 1 second of additional time to the original value.
+    Prior to backoff version 1.2, this was the default jitter behavior.
+
+    Parameters
+    ----------
+    value : float
+        The original, unjittered backoff value.
+    """
+    return value + random.random()
+
+
+def full_jitter(value: float) -> float:
+    """Jitter the value across the full range (0 to value).
+
+    This corresponds to the "Full Jitter" algorithm specified in the
+    AWS blog post on the performance of various jitter algorithms.
+    See: http://www.awsarchitectureblog.com/2015/03/backoff.html
+
+    Parameters
+    ----------
+    value : float
+        The original, unjittered backoff value.
+    """
+    return random.uniform(0, value)
+
+
+@dataclass
+class Details:
+    """Details for event handlers in SafeInvoker."""
+
+    func: Callable[[Any], Any]
+    args: Tuple[Any, ...]
+    kwargs: Dict[str, Any]
+    tries: int
+    elapsed: float
+    exception: Optional[Exception] = None
+    wait: Optional[float] = None
+
+
+# pylint: disable-next=too-many-instance-attributes
+class SafeInvoker:
+    """Wrapper class for backoff and retry triggered by exceptions.
+
+    Parameters
+    ----------
+    wait: Generator[float, None, None]
+        A generator yielding successive wait times in seconds. If the generator
+        is finite, the giveup event will be triggered when the generator raises
+        `StopIteration`.
+    exception:  Union[Type[Exception], Iterable[Type[Exception]]]
+        An exception type (or iterable of types) that triggers backoff.
+    max_tries: Optional[int] (default: None)
```

Comment:
Let's pick a safer default here, maybe just `1`? If a caller forgets to set this arg, it won't hang indefinitely here.

Reply:
typically, we need to set either `max_tries` or `max_time`. We can definitely change the default value of `max_tries` to 1, but we will have to manually set `max_tries` back to `None` when we just want to set `max_time`. And I think the `SafeInvoker` class is actually some "internal" class that average flower users won't use.

IMO, we can highlight in the docstring of `SafeInvoker` that it is recommended to set `max_tries` or `max_time` and also provide a usage example as you mentioned in one of the above reviews.

Reply:
If we recommend to set them, and also do not expect Flower users to use this, should we just remove the defaults and force the user to set these manually (and, by extension, think about appropriate values more carefully instead of just relying on defaults)?

Reply:
if removing default values for both `max_tries` and `max_time`, we may have to set both of them every time we instantiate a `Retrier`.

Alternatively, we can throw a warning in `__init__` and `invoke` methods of `Retrier` if both `max_tries` and `max_time` are set to None. How do you find this solution?

Reply:
Having to set them every time we instantiate a `Retrier` could be considered a feature - it forces the user to think about appropriate values for the situation at hand. Another benefit is that when reading the code, you can see exactly the type retry logic that is used without having to look up the defaults. I'd prefer that over raising a warning.

Reply:
Make sense. I've removed the default values for `max_tries` and `max_time`.